src/blockchain.js

```diff
@@ -1,3 +1,131 @@
+
+function reduce(arr, memo, iteratee, cb) {
+  if (typeof cb !== 'function') {
+    if (typeof memo === 'function' && typeof iteratee === 'function') {
+      cb = iteratee;
+      iteratee = memo;
+      memo = [];
+    } else {
+      throw new TypeError('expected callback to be a function');
+    }
+  }
+
+  if (!Array.isArray(arr)) {
+    cb(new TypeError('expected an array'));
+    return;
+  }
+
+  if (typeof iteratee !== 'function') {
+    cb(new TypeError('expected iteratee to be a function'));
+    return;
+  }
+
+  (function next(i, acc) {
+    if (i === arr.length) {
+      cb(null, acc);
+      return;
+    }
+
+    iteratee(acc, arr[i], function(err, val) {
+      if (err) {
+        cb(err);
+        return;
+      }
+      next(i + 1, val);
+    });
+  })(0, memo);
+};
+
+function isNewWeb3_1() {
+  return (typeof(web3.version) === "string");
+};
+
+function getAccounts(cb) {
+  if (isNewWeb3_1()) {
+    web3.eth.getAccounts().then(function(accounts) {
+      cb(null, accounts);
+      return null;
+    }).catch(function(err) {
+      cb(err);
+      return null;
+    });
+    return;
+  }
+  web3.eth.getAccounts(cb);
+};
+
+let Blockchain = {};
+
+Blockchain.connect = function(connectionList, opts, doneCb) {
+  const self = this;
+  this.web3 = null;
+  this.doFirst(function(cb) {
+    reduce(connectionList, function(prev, value, next) {
+      if (prev === false) {
+        return next(null, false);
+      }
+
+      if (value === '$WEB3' && (typeof web3 !== 'undefined' && typeof Web3 !== 'undefined')) {
+        web3.setProvider(web3.givenProvider);
+      } else if (value !== '$WEB3' && (typeof Web3 !== 'undefined' && ((typeof web3 === 'undefined') || (typeof web3 !== 'undefined' && (!web3.isConnected || (web3.isConnected && !web3.isConnected())))))) {
+        if (value.indexOf('ws://') >= 0) {
+          web3.setProvider(new Web3.providers.WebsocketProvider(value));
+        } else {
+          web3.setProvider(new Web3.providers.HttpProvider(value));
+        }
+      } else if (value === '$WEB3') {
+        return next(null, '');
+      }
+
+      getAccounts(function(err, account) {
```

Comment:
Why do you getAccounts here, if you do it in the next function anyway? Maybe I just don't understand the function flow.

Reply:
basically it's a way to verify the connection status... it's stupid but there was no other method, at least at the time this was written